process/tags_v3.go

```diff
@@ -0,0 +1,204 @@
+//	- Indexes were not deterministic: same tags in different order would produce different indexes
+// V3 sacrifices encoding performance for smaller buffer size and faster decoding.
+// V3 also guarantees the decoded tags are always sorted, which is useful for many applications.
+// V3 preserves V2's format, hence, we rely on the same decoding logic as V2.
+type V3TagEncoder struct {
+	// tags are stored as a map of tag name to position in the buffer
+	tags map[string]uint32
+	// order is the list of tags in the order they were added
+	order []string
+	// footer is the encoded footer that contains the tag positions
+	footer []byte
+	// tagPosition is the current position in the buffer where the next tag will be written
+	tagPosition uint32
+
+	// tagSetCache is a cache of previously seen tag sets to avoid re-encoding the same set
+	// The key is a hash of the sorted tags, and the value is the index in the footer
+	// This allows us to return the same index for the same set of tags.
+	tagSetCache map[uint64]int
+
+	// internal buffer for reuse
+	bufInt32 [4]byte
+	bufInt16 [2]byte
+}
+
+var v3FooterPool = sync.Pool{
+	New: func() interface{} {
+		var footer []byte
+		return &footer
+	},
+}
+
+var v3OrderPool = sync.Pool{
+	New: func() interface{} {
+		var order []string
+		return &order
+	},
+}
+
+var v3TagsPool = sync.Pool{
+	New: func() interface{} {
+		tags := make(map[string]uint32)
+		return &tags
+	},
+}
+
+var v3TagSetCachePool = sync.Pool{
+	New: func() interface{} {
+		cache := make(map[uint64]int)
+		return &cache
+	},
+}
+
+// NewV3TagEncoder creates a new V3TagEncoder with pre-allocated buffers and pools for reusing memory to reduce allocations.
+func NewV3TagEncoder() TagEncoder {
+	footer := *v3FooterPool.Get().(*[]byte)
+	order := *v3OrderPool.Get().(*[]string)
+	tags := *v3TagsPool.Get().(*map[string]uint32)
+	cache := *v3TagSetCachePool.Get().(*map[uint64]int)
+
+	return &V3TagEncoder{
+		tags:        tags,
+		order:       order[:0],
+		footer:      footer[:0],
+		tagPosition: v2PreambleLength,
+		tagSetCache: cache,
+		bufInt32:    [4]byte{},
+		bufInt16:    [2]byte{},
+	}
+}
+
+// Buffer returns the encoded buffer with the tags and footer.
+// The buffer format is:
+// - 1 byte version (3 for V3)
+// - 4 bytes footer position (where the footer starts in the buffer)
+// - N tags, each with a 2-byte length followed by the tag bytes
+// - Footer: 2 bytes tag count, followed by 4 bytes per tag (positions
+//   in the buffer where the tag starts)
+func (t *V3TagEncoder) Buffer() []byte {
+	tagsSize := t.tagPosition - v2PreambleLength
+	bufferSize := int(v2PreambleLength+tagsSize) + len(t.footer)
+	buffer := make([]byte, bufferSize)
+	pos := 0
+
+	// Write version
+	buffer[pos] = 3
+	pos++
+
+	// Write footer position
+	binary.LittleEndian.PutUint32(buffer[pos:], t.tagPosition)
+	pos += 4
+
+	// Write tags
+	for _, tag := range t.order {
+		binary.LittleEndian.PutUint16(buffer[pos:], uint16(len(tag)))
+		pos += 2
```

Comment:
Should we add a comment to explain that these two bits are to store the length of the tag?